### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by cucumberjim**
Initial cmdlets to set up a Rundeck session and do some basic job management.

### CODE REVIEW

File: RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1 [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    .DESCRIPTION
+    The `Disable-RundeckJobSchedule` function disables scheduling for a Rundeck job.
+
+    .EXAMPLE
+    Disable-RundeckJobSchedule -ID 'b090d4c8-585c-4330-8bc6-ad4783089dfd'
+
+    Demonstrates how to disable a job.
+    #>
+    param(
+        [Parameter(Mandatory,ValueFromPipelineByPropertyName)]
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1]: nit: our standard for parameters is now:

```powershell
# Parameter documentation.
[Parameter()]
[Type] $Name
```

e.g

```powershell
# The job ID.
[Parameter(Mandatory, ValueFromPipelineByPropertyName)]
[Guid] $Id
```

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    .DESCRIPTION
+    The `Get-RundeckJob` function gets a job from Rundeck.
+
+    .EXAMPLE
+    GEt-RundeckJob -ProjectName $refProject.Name -ID 'b090d4c8-585c-4330-8bc6-ad4783089dfd'
+
+    Demonstrates how to get a specific job usng its ID.
+    #>
+    param(
+        [Parameter(Mandatory,ValueFromPipeline,ValueFromPipelineByPropertyName,ParameterSetName='ByName')]
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Get-RundeckJob.ps1]: nit: Separate `Parameter` attribute properties with a comma *and* a space.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            return ($allJobs | Where-Object { $_.name -eq $Name })
+        }
+        elseif ($ID)
+        {
+            return ($allJobs | Where-Object { $_.id -eq $ID })
+        }
+        else
+        {
+            $queryFilter = [ScriptBlock]::Create($Filter)
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Get-RundeckJob.ps1]: Maybe change the type of the `$Filter` paramete to be [scriptblock]?

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            return ($allJobs | Where-Object { $_.name -like $queryFilter })
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Get-RundeckJob.ps1]: If `$queryFilter` is a script block, I believe you should be doing `($allJobs | Where-Object -FilterScript $queryFilter)`?

File: RundeckAutomation/Functions/Get-RundeckJobExecution.ps1 [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+{
+    <#
+    .SYNOPSIS
+    Gets a job job execution status from Rundeck.
+
+    .DESCRIPTION
+    The `Get-RundeckJobExecution` function gets a job execution status from Rundeck.
+
+    .EXAMPLE
+    Get-RundeckJobExecution -ProjectName $refProject.Name -ID 'b090d4c8-585c-4330-8bc6-ad4783089dfd'
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Get-RundeckJobExecution.ps1]: Example params don't match the function's signature.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        [string]
+        # The job name filter.
+        $Filter,
+
+        [Parameter(Mandatory,ValueFromPipeline,ValueFromPipelineByPropertyName,ParameterSetName='ByID')]
+        [guid]
+        # The job's ID.
+        $ID,
+
+        [Parameter(Mandatory,ValueFromPipeline,ValueFromPipelineByPropertyName)]
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Get-RundeckJob.ps1]: nit: we usually put mandatory parameters for all parameter sets as the first parameter in the list.

File: RundeckAutomation/Functions/Get-RundeckProject.ps1 [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        if ($Name)
+        {
+            $project = Invoke-RundeckRestMethod -Method 'GET' -ResourcePath "project/$($Name)"
+            return $project
+        }
+        else
+        {
+            $allProjects = Invoke-RundeckRestMethod -Method 'GET' -ResourcePath "projects"
+            $queryFilter = [ScriptBlock]::Create($Filter)
+            return ($allProjects | Where-Object { $_.name -like $queryFilter })
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Get-RundeckProject.ps1:48]: I think the `$Filter` parameter's type should be changed to `[scriptblock]` which then gets passed to `Where-Object` cmdlet's `FilterScript` parameter:

```powershell
return ($allProjects | Where-Object -FilterScript $Filter)
```

File: RundeckAutomation/Functions/Import-RundeckJob.ps1 [hunk 8]
```diff
@@ -0,0 +1,45 @@
+function Import-RundeckJob
+{
+    <#
+    .SYNOPSIS
+    Imports a job to Rundeck from an XML file.
+
+    .DESCRIPTION
+    The `Import-RundeckJob` function starts a job from Rundeck.
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Import-RundeckJob.ps1:8]: Is this description correct?

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 [hunk 9]
```diff
@@ -0,0 +1,104 @@
+
+function Invoke-RundeckRestMethod
+{
+    <#
+    .SYNOPSIS
+    Calls a method in the Rundeck REST API.
+
+    .DESCRIPTION
+    The `Invoke-RundeckRestMethod` function calls a method in the Rundeck REST API. Pass a session/connection object to the `Session` parameter (use `New-RundeckSession` to create a session object), the HTTP method to use to the `Method` parameter, the relative path to the endpoint to the `ResourcePath` parameter (i.e. everything after `api/v2/` in the endpoint's path), and the body of the request (if any) to the `Body` parameter. A result object is returned, which is different for each endpoint.
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:9]: nit: break all lines longer than 120 characters into multiple lines. In Visual Studio Code, you can setup a ruler at 120 as a reminder and guide.

Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:9]: I would rephrase the last sentence to "Any objects returned by the API endpont are returned."

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+{
+    <#
+    .SYNOPSIS
+    Calls a method in the Rundeck REST API.
+
+    .DESCRIPTION
+    The `Invoke-RundeckRestMethod` function calls a method in the Rundeck REST API. Pass a session/connection object to the `Session` parameter (use `New-RundeckSession` to create a session object), the HTTP method to use to the `Method` parameter, the relative path to the endpoint to the `ResourcePath` parameter (i.e. everything after `api/v2/` in the endpoint's path), and the body of the request (if any) to the `Body` parameter. A result object is returned, which is different for each endpoint.
+
+    .EXAMPLE
+    Invoke-RundeckRestMethod -Method Get -ResourcePath ('orders/{0}/' -f $ID)
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:12]: Nit: we now prefer expressions embedded in strings, e.g. `"orders/$($ID)"`.

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        # The body of the request.
+        $Body,
+
+        [string]
+        # URI Parameters to pass.
+        $QueryString,
+
+        [switch]
+        # Content is XML format rather than JSON.
+        $ContentIsXML
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1]: I might rename this to `$BodyXml`, and rename $Body to `$BodyJson` and make the two body parameters mutually exclusive parameter sets, with `$BodyJson` parameter set being default?

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 (hunk 9 above)
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:9]: There's no session parameter.

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+    if (-not ($_RundeckSession))
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:55]: My personal preference is to require a session object in every command. If you have a global, internal session, users can only connect to a single Rundeck session per PowerShell instance, or they have to use `New-RundeckSession` to switch back and forth. This is the pattern that PowerShell's own cmdlets follow: user creates a session/connection, then pass the session to the cmdlet that needs one.

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    $relativeUri = $ResourcePath
+    if( $QueryString )
+    {
+        $relativeUri = '{0}?{1}' -f $relativeUri,$QueryString
+    }
+    Write-Verbose "Base $($_RundeckSession.Uri.GetType()) $($_RundeckSession.Uri)"
+    Write-Verbose "Relative $($relativeUri.GetType()) $($relativeUri)"
+    $endpointUri = New-Object 'Uri' -ArgumentList @($_RundeckSession.Uri,$relativeUri)
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:68]: You can now use object constructors: `[Uri]::New($_RundeckSession.Uri, $relativeUri)` .

File: RundeckAutomation/Functions/New-RundeckProject.ps1 [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    New-RundeckProject -File '~/xmlProjectDefinition.xml'
+
+    Demonstrates how to create a project using a XML document file.
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory,ValueFromPipeline,ValueFromPipelineByPropertyName,ParameterSetName='ByJson')]
+        [string]
+        # The project's configuration in JSON format.
+        $Config,
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/New-RundeckProject.ps1]: I would probably add a mandatory `$Format` parameter, whose value must be one of `Json` or `Xml`, then rename the `$File` parameter to `$Path`. That way, users can import from a json or xml file, or use raw JSON or XML strings. Depending on the value of the `$Format` parameter, call the appropriate method on `Invoke-RundeckRestMethod`.

File: RundeckAutomation/Functions/New-RundeckSession.ps1 [hunk 15]
```diff
@@ -0,0 +1,64 @@
+function New-RundeckSession
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/New-RundeckSession.ps1:1]: If you're going to keep the global, internal session, rename this function to `Set-RundeckSession`. The `New` verb usually creates a thing and returns it.

File: RundeckAutomation/Functions/New-RundeckSession.ps1 [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+        [Parameter(Mandatory, ParameterSetName = 'username')]
+        [pscredential]
+        # The credentials to use.
+        $Credential,
+
+        [Parameter(Mandatory, ParameterSetName = 'apitoken')]
+        [string]
+        # The credentials to use.
+        $ApiToken,
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/New-RundeckSession.ps1]: Can you allow both a credential and an API token?

Another advantage to returning a session object instead of using a global object is users can mix calls to endpoints that require/use credentials with calls to endpoints that require/use the API token without needing to call `Set-RundeckSession` again. Each function would tell `Invoke-RundeckRestMethod` to use credentials or an API key.

File: RundeckAutomation/Functions/Start-RundeckJob.ps1 [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+        [switch]
+        # Wait for job to complete or fail.
+        $Wait
+    )
+
+    process
+    {
+        $jobRun = Invoke-RundeckRestMethod -Method 'POST' -ResourcePath "job/$($ID)/executions"
+        Start-Sleep -Seconds 1
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Start-RundeckJob.ps1]: What state are you waiting for? Can you call the API to find out if the job is in that state instead of sleeping?

File: RundeckAutomation/Functions/Start-RundeckJob.ps1 [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+        if ($Wait)
+        {
+            while ($jobRun.status -eq 'running')
+            {
+                Start-Sleep -Seconds 10
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Start-RundeckJob.ps1]: I would make this sleep duration a parameter.

```powershell
[TimeSpan] $WaitInterval = [TimeSpan]'00:00:10'

# ...snip...

Start-Sleep -Milliseconds $WaitInterval.TotalMilliseconds
```

File: RundeckAutomation/LICENSE [hunk 19]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (splatteredbits) [inline RundeckAutomation/LICENSE]: Usually this file goes in the repo root and we use a `CopyFile` build task in whiskey.yml to copy it into the module root before packaging. That should probably be in the default whiskey.yml file.

Same with the `NOTICE` and `README.md` files: keep them in the root, and copy them into the module root before packaging.

Also, the NOTICE file is empty.

File: RundeckAutomation/README.md [hunk 20]
```diff
@@ -0,0 +1,26 @@
+# Overview
```
Comment (splatteredbits) [inline RundeckAutomation/README.md]: Remove this file.

File: RundeckAutomation/RundeckAutomation.psd1 [hunk 21]
```diff
@@ -116,8 +128,7 @@
             Prerelease = ''
 
             # ReleaseNotes of this module
-            ReleaseNotes = @'
-'@
+            ReleaseNotes = 'https://github.com/webmd-health-services/RundeckAutomation/CHANGELOG.md'
```
Comment (splatteredbits) [inline RundeckAutomation/RundeckAutomation.psd1]: Correct URL is `https://github.com/webmd-health-services/RundeckAutomation/blob/main/CHANGELOG.md` .

File: whiskey.yml [hunk 22]
```diff
@@ -14,7 +19,7 @@ Build:
 - Exec:
       OnlyBy: BuildServer
       Path: appveyor
-      Argument: [ UpdateBuild, -Version, $(WHISKEY_SEMVER2) ]
+      Argument: [ UpdateBuild, -Version, "$(WHISKEY_SEMVER2)+$(WHISKEY_BUILD_NUMBER)" ]
 
 # Dot-sourcing files is expensive. Move all functions into .psm1 file to improve import speed. Do this before testing
```
Comment (splatteredbits) [inline whiskey.yml:25]: Add this task after the existing `MergeFile` task:

```powershell
- CopyFile:
    Path:
    - LICENSE
    - NOTICE
    - README.md
    - CHANGELOG.md
    DestinationDirectory: RundeckAutomation
```

**Review by splatteredbits — CHANGES_REQUESTED** [review]
It would be nice to have some tests for this module. [It looks like you can use Vagrant on AppVeyor](https://github.com/jonashackt/vagrant-ansible-on-appveyor), we just have to ask AppVeyor to turn on nested virtualization. Let me know if you'd like me to do that.

If you can get Rundeck running in a container, that would work on AppVeyor.

I usually write an init.ps1 script that installs the software I'm testing and add that to my build script.

**Comment by splatteredbits** [comment]
Here's the content for the NOTICE file:

```
RundeckAutomation

Copyright 2022 WebMD Health Services

This product includes software developed at WebMD Health Services (https://www.webmdhealthservices.com/).
```

File: whiskey.yml (hunk 22 above)
Comment (splatteredbits) [reply whiskey.yml:25]: Which means you'll want to add these lines to your .gitignore, too:

```
/RundeckAutomation/CHANGELOG.md
/RundeckAutomation/LICENSE
/RundeckAutomation/NOTICE
/RundeckAutomation/README.md
```

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 3 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: I wanted to replicate the filter parameter from Get-ChildItem so passing in a string that filters like -Filter "*test*" to get all jobs with test in them.  I wasn't able to get inline type transformation to work so this was my workaround.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 4 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: This was a method I found I'll test with -FilterScript, but the intent here was to pass a string filter.

**Comment by cucumberjim** [comment]
I'm ready for a re-review.  I think I've updated or replied to everything and the build looks green.

File: .gitignore [hunk 23]
```diff
@@ -1,3 +1,9 @@
 /.output
 /PSModules
-/.vscode
\ No newline at end of file
+/.vscode
```
Comment (splatteredbits) [inline .gitignore]: I think you can collapse `/.output`, `.vscode`, and `/.vargrant` to `.*` so Git just ignores all "hidden" files/directories.

File: CHANGELOG.md [hunk 24]
```diff
@@ -0,0 +1,13 @@
+# 0.2.0
```
Comment (splatteredbits) [inline CHANGELOG.md]: I think you can collapse all this into `0.1.0` since we haven't released any versions yet.

File: RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1 [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+
+    .EXAMPLE
+    Disable-RundeckJobSchedule -ID 'b090d4c8-585c-4330-8bc6-ad4783089dfd'
+
+    Demonstrates how to disable a job.
+    #>
+    param(
+        # The job ID.    
+        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
+        [GUID]
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1]: Nit a parameter's type and name should be on the same line and use the same case as the underlying type. e.g.

```powershell
[Guid] $ID
```

**Review by splatteredbits — COMMENTED** [review]
Can you take a look at my previous comments that have questions and respond to them?

File: .gitignore (hunk 23 above)
Comment (cucumberjim) [reply .gitignore]: Is `.github` needed in VCS, though?

File: RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1 (hunk 1 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1]: These should all be to the new standard now.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 2 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: These should all be to the new standard now.  I also cleaned up some of the properties that had floated in from copy/paste.

File: RundeckAutomation/Functions/Get-RundeckJobExecution.ps1 (hunk 5 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJobExecution.ps1]: Should be fixed in the next commit.

File: RundeckAutomation/Functions/Start-RundeckJob.ps1 (hunk 18 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Start-RundeckJob.ps1]: Added as a function parameter.

File: RundeckAutomation/LICENSE (hunk 19 above)
Comment (cucumberjim) [reply RundeckAutomation/LICENSE]: These should be in the right locations now and populated.

File: RundeckAutomation/README.md (hunk 20 above)
Comment (cucumberjim) [reply RundeckAutomation/README.md]: Removed.

File: RundeckAutomation/RundeckAutomation.psd1 (hunk 21 above)
Comment (cucumberjim) [reply RundeckAutomation/RundeckAutomation.psd1]: Corrected this.

File: whiskey.yml (hunk 22 above)
Comment (cucumberjim) [reply whiskey.yml:25]: Updated whiskey and the file locations.

File: CHANGELOG.md (hunk 24 above)
Comment (cucumberjim) [reply CHANGELOG.md]: Done.  I wasn't sure if intermediate commits were being published to a pre-release repo somewhere so thought I needed to update the version as I went.  I'll note the behavior for future.

File: .gitignore (hunk 23 above)
Comment (cucumberjim) [reply .gitignore]: Updated, but kept .github from being filtered out.

File: RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1 (hunk 25 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Disable-RundeckJobSchedule.ps1]: These should all be updated in the next commit.

File: .gitignore (hunk 23 above)
Comment (splatteredbits) [reply .gitignore]: `.gitignore` only prevents `git status` from showing items and `git remove` from deleting things. You can still `git add` something that matches an ignore pattern and Git will happily add it. Once a file is added to Git, the .gitignore rules never apply.

File: CHANGELOG.md (hunk 24 above)
Comment (splatteredbits) [reply CHANGELOG.md]: No automatic publishing/deploying. There's a "Deploy" button in each build job that lets you deploy to GitHub and the PowerShell Gallery.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 6 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: I think these have been updated everywhere.

File: RundeckAutomation/Functions/Get-RundeckProject.ps1 (hunk 7 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckProject.ps1:48]: The intent here was to pass a string filter like in Get-ChildItem.

File: RundeckAutomation/Functions/Import-RundeckJob.ps1 (hunk 8 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Import-RundeckJob.ps1:8]: Fixing this now.

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 (hunk 9 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:9]: Fixing this

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 (hunk 10 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:12]: Should be fixed in the next push.

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 (hunk 9 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:9]: Should be fixed in the next push.

Comment (cucumberjim) [reply RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1:9]: Should be fixed in the next push

File: RundeckAutomation/Functions/New-RundeckProject.ps1 (hunk 14 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/New-RundeckProject.ps1]: Updated in the next push.

File: RundeckAutomation/Functions/New-RundeckSession.ps1 (hunk 16 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/New-RundeckSession.ps1]: All calls should/have been able to support both.  I had a syntax error that led me to believe otherwise for one of the endpoints.  Thinking on a compromise.  I like setting it at the session level (like VMWare's PowerCLI supports), but see the value in being able to specify.

File: RundeckAutomation/Functions/Start-RundeckJob.ps1 (hunk 17 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Start-RundeckJob.ps1]: I think I was seeing a brief pause (sometimes) that has the execution queued but not accepted yet.  I'll do some more testing to see if this is actually needed or if there is a better way to handle this.

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 (hunk 11 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1]: If it is a GET, there wouldn't be a body and I was using that switch to set both the content type and the content accepted (so far they've been the same on both request and response for the call).  Thinking on this.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 4 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: I tested `-FilterScript` and it didn't work as expected transforming the string value.

**Comment by cucumberjim** [comment]
Sorry this took so long to get back to, but I think I've responded to the issues and am ready for a review.

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 4 above)
Comment (splatteredbits) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: Why are you converting `$Filter` into a script block? Why aren't you `return ($allJobs | Where-Object 'name' -Like $Filter)`?

File: RundeckAutomation/Functions/Export-RundeckProject.ps1 [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+        [String] $Name
+    )
+
+    process
+    {
+
+        $relativeUri = "project/$($Name)/export"
+        $relativeUri = '{0}?{1}' -f $relativeUri, "exportConfigs=true&exportAll=false"
+        $endpointUri = New-Object 'Uri' -ArgumentList @($_RundeckSession.Uri, $relativeUri)
+        Invoke-WebRequest -ErrorAction 'Stop' -WebSession $_RundeckSession.WebSession -Method 'GET' -Uri $endpointUri -OutFile $Path
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Export-RundeckProject.ps1]: Update to use `Invoke-RundeckRestMethod` instead of `Invoke-WebRequest`. Keep all the logic for communicating with Rundeck in one function.

**Comment by splatteredbits** [comment]
I can't stress strongly enough that you should pass a mandatory session object to each function instead of having a global session object. I know many modules do it that way, but we should be following the pattern of PowerShell itself, not third-party modules. PowerShell has you create a session to something, then pass that session object around. Global variables are in general frowned upon and not recommended.

Update `New-RundeckSession` to return a session object and update each function to have a mandatory first parameter:

```powershell
   [Parameter(Mandatory)]
   [Object] $Session,
```

File: RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1 [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+        [String] $ResourcePath,
+
+        # The body of the request.
+        [String] $Body,
+
+        # URI Parameters to pass.
+        [String] $QueryString,
+
+        # Content is XML format rather than JSON.
+        [Switch] $ContentIsXML
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Invoke-RundeckRestMethod.ps1]: Rename to `BodyIsXml`.

File: RundeckAutomation/Functions/Start-RundeckJob.ps1 [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+            $body = @{ 'runAtTime' = $dateCodeISO8601 } | ConvertTo-Json
+            $jobRun = Invoke-RundeckRestMethod -Method 'POST' -ResourcePath "job/$($ID)/executions" -Body $body
+        }
+        else
+        {
+            $jobRun = Invoke-RundeckRestMethod -Method 'POST' -ResourcePath "job/$($ID)/executions"
+        }
+
+        
+        Start-Sleep -Seconds 1
```
Comment (splatteredbits) [inline RundeckAutomation/Functions/Start-RundeckJob.ps1]: I would remove this `Start-Sleep` in favor of the `Wait` behavior below.

File: RundeckAutomation/RundeckAutomation.psd1 [hunk 29]
```diff
@@ -18,7 +18,7 @@
     RootModule = 'RundeckAutomation.psm1'
 
     # Version number of this module.
-    ModuleVersion = '0.0.0'
+    ModuleVersion = '0.1.1'
```
Comment (splatteredbits) [inline RundeckAutomation/RundeckAutomation.psd1]: Set the version to `1.0.0` and update the whiskey.yml so you publish prerelease versions.

File: Tests/Enable-RundeckJobSchedule.Tests.ps1 [hunk 30]
```diff
@@ -0,0 +1,43 @@
+
+#Requires -Version 5.1
+Set-StrictMode -Version 'Latest'
+
+Describe 'Enable-RundeckJobSchedule' {
+
+    BeforeAll {
```
Comment (splatteredbits) [inline Tests/Enable-RundeckJobSchedule.Tests.ps1]: Lint: usually `BeforeAll` blocks are outside `Describe` blocks.

File: init.ps1 [hunk 31]
```diff
@@ -16,6 +15,306 @@ Demonstrates how to call this script.
...
+
+    return @{stdout = $stdout; stderr = $stderr}
+}
+
 Set-StrictMode -Version 'Latest'
 $ErrorActionPreference = 'Stop'
 $InformationPreference = 'Continue'
+$VerbosePreference = 'Continue'
+
+Write-Information -InformationAction Continue -MessageData 'Starting init.ps1 script'
```
Comment (splatteredbits) [inline init.ps1]: Instead of all the `-InformationAction Continue` statements, add this to the top of the script:

```powershell
$InformationPreference = 'Continue'
```

File: whiskey.yml [hunk 32]
```diff
@@ -1,9 +1,14 @@
...
+    ScriptBlock: |
+        if( -not (Get-Module -Name 'Prism' -ListAvailable) )
+        {
+            Find-Module -Name 'Prism' | Select-Object -First 1 | Install-Module -Force
+        }
+        prism install | Format-Table -Auto
+
 - Version:
     Path: RundeckAutomation\RundeckAutomation.psd1
     Prerelease:
```
Comment (splatteredbits) [inline whiskey.yml:14]: Standard is now:

```yaml
- Version:
    Path: RundeckAuomation\RundeckAutomation.psd1
    Prerelease:
    - main: ""
    - "*": rc1
    IncrementPrereleaseVersion: true
```

**Review by splatteredbits — CHANGES_REQUESTED** [review]

File: RundeckAutomation/Functions/Get-RundeckJob.ps1 (hunk 4 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Get-RundeckJob.ps1]: When I supply it as a string it wasn't doing pattern matching for me.  I'll retest.

File: RundeckAutomation/Functions/Start-RundeckJob.ps1 (hunk 28 above)
Comment (cucumberjim) [reply RundeckAutomation/Functions/Start-RundeckJob.ps1]: I think I ran into a race condition where it didn't have a status on the job, but re-testing.
